lib/puppet/parser/functions/strftime.rb

```diff
@@ -0,0 +1,160 @@
+
+**ISO 8601 week-based year and week number:**
+The first week of YYYY starts with a Monday and includes YYYY-01-04.
+The days in the year before the first week are in the last week of
+the previous year.
+| Format | Meaning |
+| ------ | ------- |
+| G | The week-based year |
+| g | The last 2 digits of the week-based year (00..99) |
+| V | Week number of the week-based year (01..53) |
+
+**Week number:**
+The first week of YYYY that starts with a Sunday or Monday (according to %U
+or %W). The days in the year before the first week are in week 0.
+| Format | Meaning |
+| ------ | ------- |
+| U | Week number of the year. The week starts with Sunday. (00..53) |
+| W | Week number of the year. The week starts with Monday. (00..53) |
+
+**Seconds since the Epoch:**
+| s | Number of seconds since 1970-01-01 00:00:00 UTC. |
+
+**Literal string:**
+| Format | Meaning |
+| ------ | ------- |
+| n | Newline character (\n) |
+| t | Tab character (\t) |
+| % | Literal "%" character |
+
+**Combination:**
+| Format | Meaning |
+| ------ | ------- |
+| c | date and time (%a %b %e %T %Y) |
+| D | Date (%m/%d/%y) |
+| F | The ISO 8601 date format (%Y-%m-%d) |
+| v | VMS date (%e-%^b-%4Y) |
+| x | Same as %D |
+| X | Same as %T |
+| r | 12-hour time (%I:%M:%S %p) |
+| R | 24-hour time (%H:%M) |
+| T | 24-hour time (%H:%M:%S) |
+
+**Example**: Using `strftime` with a `Timestamp`:
+
+~~~ puppet
+$duration = Timestamp('2016-08-24T12:13:14')
+
+# Notice the duration using a format that notices <hours>:<minutes>:<seconds>
+notice($duration.strftime('%F')) # outputs '2016-08-24'
+
+# Notice the duration using a format that notices <minutes>:<seconds>
+notice($duration.strftime('%c')) # outputs 'Wed Aug 24 12:13:14 2016'
+~~~
+
+### Format directives applicable to `Timespan`:
+| Format | Meaning |
+| ------ | ------- |
+| D | Number of Days |
+| H | Hour of the day, 24-hour clock |
+| M | Minute of the hour (00..59) |
+| S | Second of the minute (00..59) |
+| L | Millisecond of the second (000..999). Digits under millisecond are truncated to not produce 1000. |
+| N | Fractional seconds digits, default is 9 digits (nanosecond). Digits under a specified length are truncated to avoid carry up |
+
+The format directive that represents the highest magnitude in the format will be allowed to overflow.
+I.e. if no "%D" is used but a "%H" is present, then the hours will be more than 23 in case the
+timespan reflects more than a day.
+
+**Example**: Using `strftime` with a Timespan and a format
+
+~~~ puppet
+$duration = Timespan({ hours => 3, minutes => 20, seconds => 30 })
+
+# Notice the duration using a format that outputs <hours>:<minutes>:<seconds>
+notice($duration.strftime('%H:%M:%S')) # outputs '03:20:30'
+
+# Notice the duration using a format that outputs <minutes>:<seconds>
+notice($duration.strftime('%M:%S')) # outputs '200:30'
+~~~
+
+- Since 4.7.0
+DOC
+) do |args|
+  function_fail(["strftime() is only available when parser/evaluator future is in effect"])
```

Comment:
We used this error messages in other 3.x functions but the message is odd given that there is only the future parser. The only way it could have been called this way is if it was loaded the 3.x way and then called via Functions, or via Scope#function_strftime.  Suggest we change this (and in all other functions using the same wording) to say "xxx() can only be called using the 4.x function API", and if we want to be helpful add "See Scope#call_function".